gulpfile.ts

```diff
@@ -0,0 +1,225 @@
+import gulp from 'gulp';
+import tap from 'gulp-tap';
+import filter from 'gulp-filter';
+import merge from 'merge-stream';
+import ts from 'gulp-typescript';
+import terser from 'gulp-terser';
+import GulpUglify from 'gulp-uglify';
+import eslint from 'gulp-eslint-new';
+import babel from 'gulp-babel';
+import { SrcOptions } from 'vinyl-fs';
+import sourcemaps from 'gulp-sourcemaps';
+import tsMacroTransformer, { macros } from 'ts-macros';
+import pathsTransformer from 'ts-transform-paths';
+import rollupTypescript from 'rollup-plugin-typescript2';
+import { CustomTransformerFactory, Program } from 'typescript';
+import minifyPrivatesTransformer from 'ts-transformer-minify-privates';
+
+/* eslint-disable @typescript-eslint/no-var-requires */
+const clean = require('gulp-clean') as () => NodeJS.ReadWriteStream;
+const rollup = require('gulp-rollup-2').rollup as (...params: unknown[]) => NodeJS.ReadWriteStream;
+/* eslint-enable @typescript-eslint/no-var-requires */
+
+function createProject(overrideSettings?: ts.Settings) {
+  return ts.createProject('tsconfig.json', {
+    getCustomTransformers: (program?: Program) => {
+      if (program === undefined) throw new Error('Program is undefined');
+      return {
+        ...pathsTransformer(program),
+        before: [
+          tsMacroTransformer(program) as unknown as CustomTransformerFactory,
+          minifyPrivatesTransformer(program)
+        ]
+      };
+    },
+    ...overrideSettings
+  });
+}
+
+function terserStream() {
+  return terser({
+    compress: false,
+    output: {
+      beautify: true
+    },
+    mangle: {
+      keep_fnames: true,
+      keep_classnames: true,
+      properties: {
+        regex: /^_private_/
+      }
+    },
+    nameCache: {}
+  });
+}
+
+function rollupStream(outFileName: string) {
+  return rollup({
+    input: 'src/index.ts',
```

Comment:
Here will make the sourcemap wrong.

Reply:
I solved it.

Reply:
Still wrong in the browser.

Because of this in the js-sdsl.min.js.map:

```javascript
{"version":3,"sources":["dist/umd/js-sdsl.js"],"names":...}
```

Reply:
![Untitled](https://user-images.githubusercontent.com/48761044/189497236-61cf1e51-f99c-46fe-9fb3-004309d38f48.png)

How about this? 
I made a source map just before uglifyjs.
```typescript
return gulp.src(`dist/umd/${output}`, { read: false, allowEmpty: true })
  .pipe(clean())
  .pipe(gulp.src(input))
  .pipe(rollupStream(sourceRoot, output))
  .pipe(terserStream())
  .pipe(sourcemap ? sourcemaps.init() : tap(() => { /* */ }))
  .pipe(GulpUglify(uglifyOpts))
  .pipe(sourcemap ? sourcemaps.write('.') : tap(() => { /* */ }))
  .pipe(gulp.dest('dist/umd'));
```

Reply:
Try to use this:

```javascript
function gulpUmdFactory(
  input: string,
  output: string,
  uglifyOpts?: GulpUglify.Options,
  sourcemap = false
) {
  macros.clear();
  let pipes = gulp.src(`dist/umd/${output}`, { read: false, allowEmpty: true })
    .pipe(clean())
    .pipe(gulp.src(input));

  function createPipes(pipes: NodeJS.ReadWriteStream) {
    return pipes
      .pipe(rollupStream(input, output))
      .pipe(terserStream()).pipe(GulpUglify(uglifyOpts));
  }

  function enableSourcemap(
    pipes: NodeJS.ReadWriteStream
  ) {
    return createPipes(pipes.pipe(sourcemaps.init()))
      .pipe(sourcemaps.mapSources(
        sourcePath => sourcePath.replace('dist/umd', '')
      ))
      .pipe(sourcemaps.write('.'));
  }

  if (sourcemap) {
    pipes = enableSourcemap(pipes);
  } else pipes = createPipes(pipes);

  return pipes.pipe(gulp.dest('dist/umd'));
}
```

Reply:
I'm so tired.

Good night~~

Reply:
It doesn't seem to work well

Reply:
I'm rewriting the rollup build script right now. Please wait

Reply:
Try this:

```javascript
function rollupStream(input: string, output: string) {
  return rollup({
    input,
    output: {
      file: output,
      format: 'umd',
      name: 'sdsl'
    },
    plugins: [
      rollupTypescript({
        tsconfigOverride: {
          compilerOptions: {
            target: 'ES5',
            module: 'ES2015',
            declaration: false
          }
        }
      })
    ]
  });
}

function gulpUmdFactory(
  input: string,
  output: string,
  uglifyOpts?: GulpUglify.Options,
  sourcemap = false
) {
  macros.clear();
  let pipes = gulp.src(`dist/umd/${output}`, { read: false, allowEmpty: true })
    .pipe(clean())
    .pipe(gulp.src(input));

  function createPipes(pipes: NodeJS.ReadWriteStream) {
    return pipes
      .pipe(rollupStream(input, output))
      .pipe(terserStream()).pipe(GulpUglify(uglifyOpts));
  }

  function enableSourcemap(
    pipes: NodeJS.ReadWriteStream
  ) {
    return createPipes(pipes.pipe(sourcemaps.init()))
      .pipe(sourcemaps.mapSources(
        sourcePath => sourcePath.replace('dist/umd/', '')
      ))
      .pipe(sourcemaps.write('.'));
  }

  if (sourcemap) {
    pipes = enableSourcemap(pipes);
  } else pipes = createPipes(pipes);

  return pipes.pipe(gulp.dest('dist/umd'));
}

gulp.task(
  'umd',
  () => gulpUmdFactory(
    'src/index.ts',
    'js-sdsl.js',
    {
      compress: false,
      output: { beautify: true },
      mangle: { keep_fnames: true }
    }
  )
);

gulp.task(
  'umd:min',
  () => gulpUmdFactory(
    'dist/umd/js-sdsl.js',
    'js-sdsl.min.js',
    {
      compress: true
    },
    true
  )
);
```

Reply:
> Try this:
> 
> ```js
> function rollupStream(input: string, output: string) {
>   return rollup({
>     input,
>     output: {
>       file: output,
>       format: 'umd',
>       name: 'sdsl'
>     },
>     context: 'this',
>     plugins: [
>       rollupTypescript({
>         tsconfigOverride: {
>           compilerOptions: {
>             target: 'ES5',
>             module: 'ES2015',
>             declaration: false
>           }
>         }
>       })
>     ]
>   });
> }
> 
> function gulpUmdFactory(
>   input: string,
>   output: string,
>   uglifyOpts?: GulpUglify.Options,
>   sourcemap = false
> ) {
>   macros.clear();
>   let pipes = gulp.src(`dist/umd/${output}`, { read: false, allowEmpty: true })
>     .pipe(clean())
>     .pipe(gulp.src(input));
> 
>   function createPipes(pipes: NodeJS.ReadWriteStream) {
>     return pipes
>       .pipe(rollupStream(input, output))
>       .pipe(terserStream()).pipe(GulpUglify(uglifyOpts));
>   }
> 
>   function enableSourcemap(pipes: NodeJS.ReadWriteStream) {
>     return createPipes(pipes.pipe(sourcemaps.init()))
>       .pipe(sourcemaps.mapSources(
>         sourcePath => sourcePath.replace('dist/umd/', '')
>       ))
>       .pipe(sourcemaps.write('.'));
>   }
> 
>   if (sourcemap) {
>     pipes = enableSourcemap(pipes);
>   } else pipes = createPipes(pipes);
> 
>   return pipes.pipe(gulp.dest('dist/umd'));
> }
> 
> gulp.task(
>   'umd',
>   () => gulpUmdFactory(
>     'src/index.ts',
>     'js-sdsl.js',
>     {
>       compress: false,
>       output: { beautify: true },
>       mangle: { keep_fnames: true }
>     }
>   )
> );
> 
> gulp.task(
>   'umd:min',
>   () => gulpUmdFactory(
>     'dist/umd/js-sdsl.js',
>     'js-sdsl.min.js',
>     {
>       compress: true
>     },
>     true
>   )
> );
> ```

The source map is created, but it doesn't seem to right.

You've been watching me so hard since yesterday, but this is my responsibility, so you don't have to struggle.
I hope you have a comfortable Mid-Autumn Festival.

Reply:
Still wrong, I got this in the map js:

```javascript
{"version":3,"file":"js-sdsl.min.js","sources":["js-sdsl.min.js"],...}
```

The source must be `js.-sdsl.js` but not `js-sdsl.min.js`.

Reply:
> You've been watching me so hard since yesterday, but this is my responsibility, so you don't have to struggle.
I hope you have a comfortable Mid-Autumn Festival.

Thanks for your concern, I'm just doing what I'm supposed to do. :D